pkg/controllers/user/clusterauthtoken/token.go

```diff
@@ -0,0 +1,181 @@
+
+	_, err := h.clusterAuthTokenLister.Get(h.namespace, token.Name)
+	if !errors.IsNotFound(err) {
+		return h.Updated(token)
+	}
+
+	err = h.updateClusterUserAttribute(token)
+	if err != nil {
+		return nil, err
+	}
+
+	clusterAuthToken, err := common.NewClusterAuthToken(token)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = h.clusterAuthToken.Create(clusterAuthToken)
+	return nil, err
+}
+
+func (h *TokenHandler) Updated(token *managementv3.Token) (runtime.Object, error) {
+	if token.ClusterName != h.clusterName {
+		return nil, nil
+	}
+
+	clusterAuthToken, err := h.clusterAuthTokenLister.Get(h.namespace, token.Name)
+	if errors.IsNotFound(err) {
+		return h.Create(token)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	err = h.updateClusterUserAttribute(token)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenEnabled := token.Enabled == nil || *token.Enabled
+	new := tokenAttributeCompare{
+		enabled:   tokenEnabled,
+		expiresAt: token.ExpiresAt,
+		username:  token.UserID,
+	}
+	old := tokenAttributeCompare{
+		enabled:   clusterAuthToken.Enabled,
+		expiresAt: clusterAuthToken.ExpiresAt,
+		username:  clusterAuthToken.UserName,
+	}
+	if reflect.DeepEqual(new, old) {
+		return nil, nil
+	}
+	clusterAuthToken.UserName = token.UserID
+	clusterAuthToken.Enabled = tokenEnabled
+	clusterAuthToken.ExpiresAt = token.ExpiresAt
+
+	_, err = h.clusterAuthToken.Update(clusterAuthToken)
+	if errors.IsNotFound(err) {
+		_, err = h.clusterAuthToken.Create(clusterAuthToken)
+	}
+	return nil, err
+}
+
+func (h *TokenHandler) Remove(token *managementv3.Token) (runtime.Object, error) {
+	if token.ClusterName != h.clusterName {
+		return nil, nil
+	}
+
+	tokens, err := h.tokenIndexer.ByIndex(tokenByUserAndClusterIndex, tokenUserClusterKey(token))
+	if errors.IsNotFound(err) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	// stale?
+	if len(tokens) == 1 {
+		lastToken := tokens[0].(*managementv3.Token)
+		if token.Name == lastToken.Name {
+			// we are about to remove the last token for this user & cluster,
+			// also remove user data from cluster
+			err := h.clusterUserAttribute.Delete(token.UserID, &metav1.DeleteOptions{})
+			if err != nil {
+				logrus.Error(err)
+			}
+		}
+	}
+	err = h.clusterAuthToken.Delete(token.Name, &metav1.DeleteOptions{})
+	if errors.IsNotFound(err) {
+		return nil, nil
+	}
+	return nil, err
+}
+
+func (h *TokenHandler) updateClusterUserAttribute(token *managementv3.Token) error {
+	userID := token.UserID
+	user, err := h.userLister.Get("", userID)
+	if err != nil {
+		return err
+	}
+
+	userAttribute, err := h.userAttributeLister.Get("", userID)
+	if err != nil {
+		return err
+	}
+
+	groups := []string{}
+	for _, gp := range userAttribute.GroupPrincipals {
+		for i := range gp.Items {
+			groups = append(groups, gp.Items[i].Name)
+		}
+	}
+	sort.Strings(groups)
+
+	userEnabled := user.Enabled == nil || *user.Enabled
+
+	clusterUserAttribute, err := h.clusterUserAttributeLister.Get(h.namespace, userID)
```

Comment:
Move the error check here. If it doesn't exist create it, if it does, do the update check.